ruby_oop/lib/scrambler.rb

```diff
@@ -0,0 +1,139 @@
+# The `#scramble` method should return itself, so that the user can rescramble it again, or use the data.
+#
+# Hints:
+#  * expose players as an attribute reader
+#  * expose activities as an attribute reader
+#  * activities should respond to `#participants` which return the players that were added to the activity
+#  * an instance of the `Scrambler` class should allow calling `Scrambler#scramble` multiple times
+
+class Participant
+  attr_reader :team
+
+  def initialize(codename:, team:)
+    @codename  = codename
+    @team      = team
+    # TODO: Maybe add @team_size for easier sort in Scrambler#players_by_team_sorted
+  end
+end
+
+class Activity
+  attr_reader   :id, :slots
+  attr_accessor :participants
+
+  def initialize(id:, slots:)
+    @id           = id
+    @slots        = slots
+    @participants = []
+  end
+
+  def slots_available?
+    @slots > @participants.size
+  end
+
+  def participants_count(team = nil)
+    return participants.size if team.nil?
+
+    participants.count { |player| player.team == team }
+  end
+end
+
+class Scrambler
+  attr_reader :players, :activities
+
+  def initialize(players:, activities:)
+    validate(players, activities)
+
+    @players    = players.map { |player| Participant.new(codename: player[:codename], team: player[:team]) }
+    @activities = activities.map { |activity| Activity.new(id: activity[:id], slots: activity[:slots]) }
+  end
+
+  def scramble
+    players_crisscrossed.each { |player| most_suitable_activity(player).participants << player }
+
+    self
+  end
+
+  # Too expensive plus using while and nested loops
+  def scramble_old
+    while activities_available? && players_available?
+      players_by_team_sorted.each_value do |players|
+        @activities.each do |activity|
+          next activity unless activity.slots_available?
+          next players  if     players.empty?
+
+          activity.participants << players.shift
+        end
+      end
+    end
+
+    self
+  end
+
+  private
+
+  def validate(players, activities)
+    raise RuntimeError unless activities.size == activities.group_by { |activity| activity[:id]    }.size
+    raise RuntimeError unless players.size    == players.group_by    { |player| player[:codename]  }.size
+    raise RuntimeError unless players.size    == activities.sum      { |activity| activity[:slots] }
+  end
+
+  def players_crisscrossed
+    players_crisscross = []
+
+    players_by_team_sorted.values.first.size.times do
+      players_by_team_sorted.each_value do |players|
+        next players if players.empty?
+
+        players_crisscross << players.shift
+      end
+    end
+
+    players_crisscross
+  end
+
+  def players_by_team_sorted
+    number_of_players_by_team = @players.map(&:team).tally
+
+    @players_by_team_sorted ||=
+      @players
+      .shuffle
+      .sort { |p1, p2| number_of_players_by_team[p2.team] <=> number_of_players_by_team[p1.team] }
+      .group_by(&:team)
```

Comment:
Nije mi bas jasno sta se tu dogadja: igraci se sortiraju po velicini time i onda se grupiraju po timu. Nije li to samo sortiranje timova po velicini?